homeassistant/components/statistics/sensor.py

```diff
@@ -487,48 +501,61 @@ def _derive_unit_of_measurement(self, new_state: State) -> str | None:
             unit = base_unit + "/sample"
         elif self._state_characteristic == STAT_CHANGE_SECOND:
             unit = base_unit + "/s"
+
         return unit
 
-    @property
-    def device_class(self) -> SensorDeviceClass | None:
-        """Return the class of this device."""
+    def _calculate_device_class(
+        self, new_state: State, unit: str | None
+    ) -> SensorDeviceClass | None:
+        """Return the calculated device class."""
+
+        # Calculate the device class based on the state characteristics,
+        # the source device class and the unit of measurement is
+        # in the device class units list.
+        device_class: SensorDeviceClass | None = None
         if self._state_characteristic in STATS_DATETIME:
             return SensorDeviceClass.TIMESTAMP
         if self._state_characteristic in STATS_NUMERIC_RETAIN_UNIT:
-            source_state = self.hass.states.get(self._source_entity_id)
-            if source_state is None:
+            device_class = new_state.attributes.get(ATTR_DEVICE_CLASS)
+            if device_class is None:
                 return None
-            source_device_class = source_state.attributes.get(ATTR_DEVICE_CLASS)
-            if source_device_class is None:
+            if (
+                sensor_device_class := try_parse_enum(SensorDeviceClass, device_class)
+            ) is None:
                 return None
-            sensor_device_class = try_parse_enum(SensorDeviceClass, source_device_class)
-            if sensor_device_class is None:
+            if (
+                sensor_device_class
+                and (
+                    sensor_state_classes := DEVICE_CLASS_STATE_CLASSES.get(
+                        sensor_device_class
+                    )
+                )
+                and sensor_state_classes
+                and SensorStateClass.MEASUREMENT not in sensor_state_classes
+            ):
                 return None
-            sensor_state_classes = DEVICE_CLASS_STATE_CLASSES.get(
-                sensor_device_class, set()
-            )
-            if SensorStateClass.MEASUREMENT not in sensor_state_classes:
+            if device_class not in DEVICE_CLASS_UNITS:
+                return None
+            if (
+                device_class in DEVICE_CLASS_UNITS
+                and unit not in DEVICE_CLASS_UNITS[device_class]
+            ):
                 return None
-            return sensor_device_class
-        return None
 
-    @property
-    def state_class(self) -> SensorStateClass | None:
-        """Return the state class of this entity."""
+        return device_class
+
+    def _calculate_state_class(self, new_state: State) -> SensorStateClass | None:
+        """Return the calculated state class."""
         if self._state_characteristic in STATS_NOT_A_NUMBER:
             return None
+        # Automatically set state class when output is a number
```

Comment:
```suggestion
    def _calculate_state_class(self, new_state: State) -> SensorStateClass | None:
        """Return the calculated state class.
        
        Will be None if the characteristics is not numerical, otherwise
        SensorStateClass.MEASUREMENT.
        """
        if self._state_characteristic in STATS_NOT_A_NUMBER:
            return None
```